test/unit/eval/typval_spec.lua

```diff
@@ -1057,39 +1057,108 @@ describe('typval.c', function()
         eq(empty_list, typvalt2lua(rettv4))
       end)
     end)
-  end)
-  describe('join()', function()
-    local function list_join(l, sep, ret)
-      local ga = ga_alloc()
-      eq(ret or OK, lib.tv_list_join(ga, l, sep))
-      if ga.ga_data == nil then return ''
-      else return ffi.string(ga.ga_data)
+    describe('join()', function()
+      local function list_join(l, sep, ret)
+        local ga = ga_alloc()
+        eq(ret or OK, lib.tv_list_join(ga, l, sep))
+        if ga.ga_data == nil then return ''
+        else return ffi.string(ga.ga_data)
+        end
       end
-    end
-    itp('works', function()
-      local l
-      l = list('boo', 'far')
-      eq('boo far', list_join(l, ' '))
-      eq('boofar', list_join(l, ''))
-
-      l = list('boo')
-      eq('boo', list_join(l, ' '))
-
-      l = list()
-      eq('', list_join(l, ' '))
-
-      l = list({}, 'far')
-      eq('{} far', list_join(l, ' '))
-
-      local recursive_list = {}
-      recursive_list[1] = recursive_list
-      l = ffi.gc(list(recursive_list, 'far'), nil)
-      eq('[[...@0]] far', list_join(l, ' '))
-
-      local recursive_l = l.lv_first.li_tv.vval.v_list
-      local recursive_li = recursive_l.lv_first
-      lib.tv_list_item_remove(recursive_l, recursive_li)
-      lib.tv_list_free(l, true)
+      itp('works', function()
+        local l
+        l = list('boo', 'far')
+        eq('boo far', list_join(l, ' '))
+        eq('boofar', list_join(l, ''))
+
+        l = list('boo')
+        eq('boo', list_join(l, ' '))
+
+        l = list()
+        eq('', list_join(l, ' '))
+
+        l = list({}, 'far')
+        eq('{} far', list_join(l, ' '))
+
+        local recursive_list = {}
+        recursive_list[1] = recursive_list
+        l = ffi.gc(list(recursive_list, 'far'), nil)
+        eq('[[...@0]] far', list_join(l, ' '))
+
+        local recursive_l = l.lv_first.li_tv.vval.v_list
+        local recursive_li = recursive_l.lv_first
+        lib.tv_list_item_remove(recursive_l, recursive_li)
+        lib.tv_list_free(l, true)
+      end)
+    end)
+    describe('equal()', function()
+      itp('compares empty and NULL lists correctly', function()
+        local l = list()
+        local l2 = list()
+
+        -- NULL lists are not equal to empty lists
+        eq(false, lib.tv_list_equal(l, nil, true, false))
+        eq(false, lib.tv_list_equal(nil, l, false, false))
+        eq(false, lib.tv_list_equal(nil, l, false, true))
+        eq(false, lib.tv_list_equal(l, nil, true, true))
+
+        -- Yet NULL lists are equal themselves
```

Comment:
No. If there is one NULL list then “NULL list is equal to itself”. If there are “many” NULL lists that for some reason can be considered different (semantically in first place, there really can only be a single NULL list) then “are equal themselves” because “different” lists are compared.

Your variant sounds like “there are many NULL lists, but we are comparing only same lists”. My variant sounds like “there are many NULL lists, we may be comparing different lists”. I like second variant more then “there is one NULL list which is compared to itself” (first variant from this comment) and consider your correction incorrect. Also because there is “As well as empty lists” below: both alternative variants make this phrase incorrect (empty lists are compared not only to themselves) and require altering it.